readme.txt
=== Gravity Flow Form Connector Extension ===
Contributors: stevehenty
Tags: gravity forms, approvals, workflow
Requires at least: 4.4
Tested up to: 5.4.2
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Create, update and link entries in Gravity Flow.

== Description ==

The Gravity Flow Form Connector Extension is an advanced extension for Gravity Flow.

Gravity Flow is a premium Add-On for [Gravity Forms](https://gravityflow.io/gravityforms)

= Requirements =

1. [Purchase and install Gravity Forms](https://gravityflow.io/gravityforms)
1. [Purchase and install Gravity Flow](https://gravityflow.io)
1. Wordpress 4.7+
1. Gravity Forms 2.1+
1. Gravity Flow 1.7+


= Support =
If you find any that needs fixing, or if you have any ideas for improvements, please get in touch:
https://gravityflow.io/contact/


== Installation ==

1.  Download the zipped file.
1.  Extract and upload the contents of the folder to /wp-contents/plugins/ folder
1.  Go to the Plugin management page of WordPress admin section and enable the 'Gravity Flow Form Connector Extension' plugin.

== Frequently Asked Questions ==

= Which license of Gravity Flow do I need? =
The Gravity Flow Form Connector Extension will work with any license of [Gravity Flow](https://gravityflow.io).


== ChangeLog ==

= 1.7.4 =
- Added the gravityflowformconnector_new_entry_form and gravityflowformconnector_update_field_values_form filters.
- Fixed issue with merge tag evaluation that caused a fatal error involving certain conditional logic setup. Update to Gravity Flow 2.5.11 will also be required.


= 1.7.3 =
- Added gravityflowformconnector_update_entry_form filter which allows the form of entry for update to be customized/hydrated.
- Updated translations. Added Catalan and Arabic.

= 1.7.2 =
- Fixed an issue where the form submission step cannot be released if the payment method set in the target form is Stripe Checkout (Stripe add-on > 3.0).
- Fixed an issue with hidden field values mapped with form submission step not being stored into child entry correctly.
- Fixed an issue with Update An Entry local/remote processing that prevents updates when the form of selected entry does not match step settings target form.

= 1.7.1 =
- Fixed an issue with the form submission step when the target page contains the submit shortcode. Requires Gravity Flow 2.5.4.
- Updated translations.

= 1.7 =
- Updated help text on Update Fields step settings screen.
- Updated translations.
- Fixed the form display and submission processes for the Form Submission step running for both the parent and nested form when used with GP Nested Forms.

= 1.6 =
- Added support for the license key constant GRAVITY_FLOW_FORM_CONNECTOR_LICENSE_KEY.
- Updated Gravity_Flow_Step_Update_Entry::get_entry_lookup_search_criteria() to support processing merge tags. Credit: Jake Jackson.
- Deprecated Gravity_Flow_Step_Update_Entry::gravityflow_entry_lookup_sort_criteria()
- Deprecated Gravity_Flow_Step_Update_Entry::gravityflow_entry_lookup_search_criteria()

= 1.5 =
- Added a new step type, Update Fields, for mapping values from other entries into the current entry including support for lookup conditional logic.
- Fixed a conflict with ForGravity's Live Population Add-On.

= 1.4.2 =
- Added support for the form submission merge tag to be used in the confirmation for email assignees.
- Removed the discussion field from the available options for mapping on the Form submissions step.

= 1.4.1 =
- Fixed an issue with New/Update an Entry steps with mapped list field types not copying data.
- Fixed an issue with New/Update an Entry steps settings for list field types. Individual columns are no longer shown for selection.


= 1.4 =
- Added the Delete Entry step.
- Added the "gravityflowformconnector_{step_type}_use_choice_text" filter allowing the choice text to be returned instead of the choice values.
- Fixed an issue where a checkbox field (selected) mapped to a text field would return the choice text instead of the choice values.
- Fixed a PHP deprecation notice on PHP 7.2 with the Form Submission step.

= 1.3 =
- Added support for the token attribute to the {workflow_form_submission_url} and {workflow_form_submission_link} merge tags.
- Added the Assignee setting to the Update Entry step to allow the assignee to be selected for User Input and Approval actions.
- Fixed a misleading message at the top of the form for email assignees when the link is not valid.
- Fixed an issue with the Form Submission step where the role and email assignees can't complete the step.
- Fixed an issue with the Form Submission step where hidden and administrative fields may not get mapped.
- Updated Members 2.0 integration to use human readable labels for the capabilities. Requires Gravity Flow 1.8.1 or greater.


= 1.2.1 =
- Added support for steps extending Gravity_Flow_Step_Form_Submission
- Fixed an issue with the Parent-Child Forms extension where an invalid link message is displayed when the parent entry is on a step that is not a Form Submission step.

= 1.2 =
- Added the Store New Entry ID setting to the New Entry step settings.
- Added the gravityflowformconnector_update_entry_id filter to allow the target entry ID to be modified.
  Example:
  add_filter( 'gravityflowformconnector_update_entry_id', 'sh_gravityflowformconnector_update_entry_id', 10, 5);
  function sh_gravityflowformconnector_update_entry_id( $target_entry_id, $target_form_id, $entry, $form, $step ) {
      // Custom search for the target entry ID based on the value of field ID 4.
      $search_criteria['status'] = 'active';
      $search_criteria['field_filters'][] = array( 'key' => '2', 'value' => $entry['4'] );
      $entries = GFAPI::get_entries( $target_form_id, $search_criteria );
      // Return the ID of the first entry in the results.
      return $entries[0]['id'];
  }
- Added support for updating the same entry when the target and source forms are the same. Select Entry ID (Self) in the Entry ID field setting.
- Fixed an issue with the approval action of the Update Entry Step for entries created with the New Entry step.
- Fixed an issue with the field mappings which may affect some forms.
- Fixed an issue with the update step when triggered by a schedule or the expiration of a previous step where the approval or user input action does not complete.
- Fixed an issue with the update step where remote approval and user input steps can fail on some servers. Requires Gravity Flow 1.6.2-dev+.


= 1.1 =
- Added translations for Chinese (China) and Dutch (Netherlands).
- Added integration with the Gravity Flow Parent-Child Forms Extension; a parent form can now be selected for the 'Entry ID Field' setting on the 'Update an Entry' step.
- Added the Form Submission step.
- Fixed an issue with the value for choice based Poll, Quiz, and Survey fields in the new or updated entry.

= 1.0.1.2 =
- Added support for mapping the created_by field in the target entry.
- Fixed a fatal error which could occur when using the 'Update an Entry' step type and the 'Entry ID Field' setting was not configured.

= 1.0.1.1 =
- Added support for merge tag processing in the custom values fields of mappings.

= 1.0.1 =
- Added support for custom values in the mapping. Requires Gravity Flow 1.3.0.10.

= 1.0.0 =
All new!


change_log.txt
- Added the gravityflowformconnector_new_entry_form and gravityflowformconnector_update_field_values_form filters.
- Fixed issue with merge tag evaluation that caused a fatal error involving certain conditional logic setup. Update to Gravity Flow 2.5.11 will also be required.


tests/bootstrap.php
<?php

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = dirname( dirname( __FILE__ ) ) . '/tmp/wordpress-tests-lib';
}

require_once $_tests_dir . '/includes/functions.php';

function _manually_load_plugin() {
	if ( ! class_exists( 'RGForms' ) ) {
		require dirname( dirname( __FILE__ ) ) . '/tmp/gravityforms/gravityforms.php';
	}
	require dirname( dirname( __FILE__ ) ) . '/tmp/gravityflow/gravityflow.php';
	require dirname( dirname( __FILE__ ) ) . '/formconnector.php';

	gf_upgrade()->install();
	GFForms::loaded();
	gravity_flow()->setup();
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

require $_tests_dir . '/includes/bootstrap.php';

$_gf_tests_dir = GFCommon::get_base_path() . '/tests';
require $_gf_tests_dir . '/gravityforms-utils.php';
require $_gf_tests_dir . '/gravityforms-testcase.php';


formconnector.php
<?php
/*
Plugin Name: Gravity Flow Form Connector
Plugin URI: https://gravityflow.io
Description: Form Connector Extension for Gravity Flow.
Version: 1.7.5-dev
Author: Gravity Flow
Author URI: https://gravityflow.io
License: GPL-2.0+

------------------------------------------------------------------------
Copyright 2015-2020 Steven Henty S.L.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License, version 2, as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

define( 'GRAVITY_FLOW_FORM_CONNECTOR_VERSION', '1.7.5-dev' );
define( 'GRAVITY_FLOW_FORM_CONNECTOR_EDD_ITEM_ID', 3375 );
define( 'GRAVITY_FLOW_FORM_CONNECTOR_EDD_ITEM_NAME', 'Form Connector' );

add_action( 'gravityflow_loaded', array( 'Gravity_Flow_Form_Connector_Bootstrap', 'load' ), 1 );

class Gravity_Flow_Form_Connector_Bootstrap {

	public static function load() {
		require_once( 'includes/class-dynamic-hook.php' );
		require_once( 'includes/class-step-form-submission.php' );
		require_once( 'includes/class-step-new-entry.php' );
		require_once( 'includes/class-step-update-entry.php' );
		require_once( 'includes/class-step-update-fields.php' );
		require_once( 'includes/class-step-delete-entry.php' );
		require_once( 'includes/class-merge-tag-form-submission.php' );
		require_once( 'class-form-connector.php' );

		// Registers the class name with GFAddOn.
		GFAddOn::register( 'Gravity_Flow_Form_Connector' );

		if ( defined( 'GRAVITY_FLOW_FORM_CONNECTOR_LICENSE_KEY' ) ) {
			gravity_flow_form_connector()->license_key = GRAVITY_FLOW_FORM_CONNECTOR_LICENSE_KEY;
		}
	}
}

function gravity_flow_form_connector() {
	if ( class_exists( 'Gravity_Flow_Form_Connector' ) ) {
		return Gravity_Flow_Form_Connector::get_instance();
	}
}


add_action( 'admin_init', 'gravityflow_form_connector_edd_plugin_updater', 0 );

function gravityflow_form_connector_edd_plugin_updater() {

	if ( ! function_exists( 'gravity_flow_form_connector' ) ) {
		return;
	}

	$gravity_flow_form_connector = gravity_flow_form_connector();
	if ( $gravity_flow_form_connector ) {

		if ( defined( 'GRAVITY_FLOW_FORM_CONNECTOR_LICENSE_KEY' ) ) {
			$license_key = GRAVITY_FLOW_FORM_CONNECTOR_LICENSE_KEY;
		} else {
			$settings = $gravity_flow_form_connector->get_app_settings();
			$license_key = trim( rgar( $settings, 'license_key' ) );
		}

		$edd_updater = new Gravity_Flow_EDD_SL_Plugin_Updater( GRAVITY_FLOW_EDD_STORE_URL, __FILE__, array(
			'version'   => GRAVITY_FLOW_FORM_CONNECTOR_VERSION,
			'license'   => $license_key,
			'item_id' => GRAVITY_FLOW_FORM_CONNECTOR_EDD_ITEM_ID,
			'author'    => 'Gravity Flow',
		) );
	}

}


includes/class-step-update-entry.php
<?php

/**
 * Gravity Flow Update Entry Step
 *
 *
 * @package     GravityFlow
 * @subpackage  Classes/Step
 * @copyright   Copyright (c) 2015-2018, Steven Henty S.L.
 * @license     http://opensource.org/licenses/gpl-3.0.php GNU Public License
 * @since       1.0
 */

if ( class_exists( 'Gravity_Flow_Step' ) ) {

	class Gravity_Flow_Step_Update_Entry extends Gravity_Flow_Step_New_Entry {
		public $_step_type = 'update_entry';

		public function get_label() {
			return esc_html__( 'Update an Entry', 'gravityflowformconnector' );
		}

		/**
		 * Returns the array of settings for this step.
		 *
		 * @return array
		 */
		public function get_settings() {

			$forms          = $this->get_forms();
			$form_choices[] = array(
				'label' => esc_html__( 'Select a Form', 'gravityflowformconnector' ),
				'value' => '',
			);
			foreach ( $forms as $form ) {
				$form_choices[] = array( 'label' => $form->title, 'value' => $form->id );
			}

			$action_choices = $this->action_choices();

			$settings = array(
				'title'  => esc_html__( 'Update an Entry', 'gravityflow' ),
				'fields' => array(
					array(
						'name'          => 'server_type',
						'label'         => esc_html__( 'Site', 'gravityflowformconnector' ),
						'type'          => 'radio',
						'default_value' => 'local',
						'horizontal'    => true,
						'onchange'      => 'jQuery(this).closest("form").submit();',
						'choices'       => array(
							array( 'label' => esc_html__( 'This site', 'gravityflowformconnector' ), 'value' => 'local' ),
							array( 'label' => esc_html__( 'A different site', 'gravityflowformconnector' ), 'value' => 'remote' ),
						),
					),
					array(
						'name'       => 'remote_site_url',
						'label'      => esc_html__( 'Site Url', 'gravityflowformconnector' ),
						'type'       => 'text',
						'dependency' => array(
							'field'  => 'server_type',
							'values' => array( 'remote' ),
						),
					),
					array(
						'name'       => 'remote_public_key',
						'label'      => esc_html__( 'Public Key', 'gravityflowformconnector' ),
						'type'       => 'text',
						'dependency' => array(
							'field'  => 'server_type',
							'values' => array( 'remote' ),
						),
					),
					array(
						'name'       => 'remote_private_key',
						'label'      => esc_html__( 'Private Key', 'gravityflowformconnector' ),
						'type'       => 'text',
						'dependency' => array(
							'field'  => 'server_type',
							'values' => array( 'remote' ),
						),
					),
					array(
						'name'     => 'target_form_id',
						'label'    => esc_html__( 'Form', 'gravityflowformconnector' ),
						'type'     => 'select',
						'onchange' => "jQuery('#action').val('update');jQuery(this).closest('form').submit();",
						'choices'  => $form_choices,
					),
					array(
						'name'       => 'action',
						'label'      => esc_html__( 'Action', 'gravityflowformconnector' ),
						'type'       => count( $action_choices ) == 1 ? 'hidden' : 'select',
						'default_value' => 'update',
						'horizontal' => true,
						'onchange'   => "jQuery(this).closest('form').submit();",
						'choices'    => $action_choices,
					),
				),
			);

			$entry_id_field = array(
				'name'       => 'update_entry_id',
				'label'      => esc_html__( 'Entry ID Field', 'gravityflowformconnector' ),
				'type'       => 'field_select',
				'tooltip'    => __( 'Select the field which will contain the entry ID of the entry that will be updated. This is used to lookup the entry so it can be updated.', 'gravityflowformconnector' ),
				'required'   => true,
				'dependency' => array(
					'field'  => 'action',
					'values' => array( 'update', 'approval', 'user_input' ),
				),
			);

			if ( function_exists( 'gravity_flow_parent_child' ) ) {
				$parent_form_choices = array();
				$entry_meta          = gravity_flow_parent_child()->get_entry_meta( array(), rgget( 'id' ) );

				foreach ( $entry_meta as $meta_key => $meta ) {
					$parent_form_choices[] = array(
						'value' => $meta_key,
						'label' => $meta['label'],
					);
				}

				if ( ! empty( $parent_form_choices ) ) {
					$entry_id_field['args']['append_choices'] = $parent_form_choices;
				}
			}

			if ( $this->get_setting( 'target_form_id' ) == $this->get_form_id() ) {
				$self_entry_id_choice = array( array( 'label' => esc_html__( 'Entry ID (Self)', 'gravityflowformconnector' ), 'value' => 'id' ) );
				if ( ! isset( $entry_id_field['args']['append_choices'] ) ) {
					$entry_id_field['args']['append_choices'] = array();
				}
				$entry_id_field['args']['append_choices'] = array_merge( $entry_id_field['args']['append_choices'], $self_entry_id_choice );
			}

			$settings['fields'][] = $entry_id_field;
			$settings['fields'][] = array(
				'name'       => 'approval_status_field',
				'label'      => esc_html__( 'Approval Status Field', 'gravityflowformconnector' ),
				'type'       => 'field_select',
				'dependency' => array(
					'field'  => 'action',
					'values' => array( 'approval' ),
				),
			);

			$mapping_field = array(
				'name'                => 'mappings',
				'label'               => esc_html__( 'Field Mapping', 'gravityflowformconnector' ),
				'type'                => 'generic_map',
				'enable_custom_key'   => false,
				'enable_custom_value' => true,
				'key_field_title'     => esc_html__( 'Field', 'gravityflowformconnector' ),
				'value_field_title'   => esc_html__( 'Value', 'gravityflowformconnector' ),
				'value_choices'       => $this->value_mappings(),
				'key_choices'         => $this->field_mappings(),
				'tooltip'             => '<h6>' . esc_html__( 'Mapping', 'gravityflowformconnector' ) . '</h6>' . esc_html__( 'Map the fields of this form to the selected form. Values from this form will be saved in the entry in the selected form', 'gravityflowformconnector' ),
				'dependency'          => array(
					'field'  => 'action',
					'values' => array( 'update', 'user_input' ),
				),
			);

			$settings['fields'][] = $mapping_field;

			$action = $this->get_setting( 'action' );

			if ( $this->get_setting( 'server_type' ) == 'remote' && in_array( $action, array(
					'approval',
					'user_input',
				) )
			) {
				$target_form_id = $this->get_setting( 'target_form_id' );
				if ( ! empty( $target_form_id ) ) {
					$settings['fields'][] = array(
						'name'    => 'remote_assignee',
						'label'   => esc_html__( 'Assignee', 'gravityflowformconnector' ),
						'type'    => 'select',
						'choices' => $this->get_remote_assignee_choices( $target_form_id ),
					);
				}
			} elseif ( $this->get_setting( 'server_type' ) == 'local' && $this->get_setting( 'action' ) == 'user_input' ) {
				$target_form_id = $this->get_setting( 'target_form_id' );
				if ( ! empty( $target_form_id ) ) {
					$settings['fields'][] = array(
						'name'    => 'local_assignee',
						'label'   => esc_html__( 'Assignee', 'gravityflowformconnector' ),
						'type'    => 'select',
						'choices' => $this->get_local_assignee_choices( $target_form_id ),
					);
				}
			}

			return $settings;
		}

		/**
		 * Returns the array of choices for the action setting.
		 *
		 * @return array
		 */
		public function action_choices() {
			$choices = array(
				array( 'label' => esc_html__( 'Update an Entry', 'gravityflow' ), 'value' => 'update' ),
			);

			$target_form_id = $this->get_setting( 'target_form_id' );

			if ( empty( $target_form_id ) ) {
				return $choices;
			}

			$has_approval_step   = false;
			$has_user_input_step = false;

			if ( $this->get_setting( 'server_type' ) == 'remote' ) {
				$steps = $this->get_remote_steps( $target_form_id );
				if ( $steps ) {
					foreach ( $steps as $step ) {
						if ( $step['type'] == 'approval' ) {
							$has_approval_step = true;
						} elseif ( $step['type'] == 'user_input' ) {
							$has_user_input_step = true;
						}
					}
				}
			} else {

				$api   = new Gravity_Flow_API( $target_form_id );
				$steps = $api->get_steps();

				foreach ( $steps as $step ) {
					if ( $step->get_type() == 'approval' ) {
						$has_approval_step = true;
					} elseif ( $step->get_type() == 'user_input' ) {
						$has_user_input_step = true;
					}
				}
			}

			if ( $has_approval_step ) {
				$choices[] = array( 'label' => esc_html__( 'Approval', 'gravityflow' ), 'value' => 'approval' );
			}
			if ( $has_user_input_step ) {
				$choices[] = array( 'label' => esc_html__( 'User Input', 'gravityflow' ), 'value' => 'user_input' );
			}

			return $choices;
		}

		/**
		 * Updates a local entry.
		 *
		 * @return bool Has the step finished?
		 */
		public function process_local_action() {
			$entry = $this->get_entry();

			$target_form_id = $this->target_form_id;

			$api = new Gravity_Flow_API( $target_form_id );

			$steps = $api->get_steps();

			$form = $this->filter_form( $this->get_form(), $entry );

			$target_entry_id = rgar( $entry, $this->update_entry_id );

			$target_entry_id = apply_filters( 'gravityflowformconnector_update_entry_id', $target_entry_id, $target_form_id, $entry, $form, $this );

			if ( empty( $target_entry_id ) ) {
				return true;
			}

			$target_entry = GFAPI::get_entry( $target_entry_id );

			if ( is_wp_error( $target_entry ) ) {
				return true;
			}

			if ( $target_form_id !== $target_entry['form_id'] ) {
				$this->log_debug( __METHOD__ . '(): The form of selected entry does not match step settings expected form. No updates performed.' );
				return true;
			}

			$new_entry = $this->do_mapping( $form, $entry );

			$new_entry['form_id'] = $this->target_form_id;

			if ( in_array( $this->action, array( 'update', 'user_input' ) ) ) {
				if ( ! is_wp_error( $target_entry ) ) {
					foreach ( $new_entry as $key => $value ) {
						$target_entry[ (string) $key ] = $value;
					}
					GFAPI::update_entry( $target_entry );
				}
			}

			if ( in_array( $this->action, array( 'approval', 'user_input' ) ) && $steps ) {

				if ( empty( $target_entry['workflow_final_status'] ) || $target_entry['workflow_final_status'] == 'pending' ) {
					$current_step = $api->get_current_step( $target_entry );

					if ( $current_step ) {

						$status = ( $this->action == 'approval' ) ? strtolower( rgar( $entry, $this->approval_status_field ) ) : 'complete';

						if ( empty( $this->local_assignee ) || $this->local_assignee == 'created_by' ) {
							$assignee_key = gravity_flow()->get_current_user_assignee_key();
							if ( ! $assignee_key && rgar( $entry, 'created_by' ) ) {
								$assignee_key = 'user_id|' . $entry['created_by'];
							}
						} else {
							$assignee_key = $this->local_assignee;
						}

						$assignees = array();

						if ( $assignee_key ) {
							$is_assignee = $current_step->is_assignee( $assignee_key );

							if ( $is_assignee ) {
								$assignee = new Gravity_Flow_Assignee( $assignee_key, $current_step );
								$assignees = array( $assignee );
							} else {
								// Assignee not set by the local_assignee setting or by current user.
								// Could be legacy settings triggered by cron or anonymous form submission.
								// Complete step for all assignees.
								$assignees = $current_step->get_assignees();
							}
						}

						$form = GFAPI::get_form( $this->target_form_id );

						$process_required = false;
						foreach ( $assignees as $assignee ) {
							$result = $current_step->process_assignee_status( $assignee, $status, $form );
							if ( $result ) {
								$process_required = true;
							}
						}

						if ( $process_required ) {
							$api->process_workflow( $target_entry_id );
						}
					}
				}
			}

			return true;
		}

		/**
		 * Updates a remote entry.
		 *
		 *
		 * @return bool Has the step finished?
		 */
		public function process_remote_action() {
			$entry = $this->get_entry();

			$form = $this->filter_form( $this->get_form(), $entry );

			$new_entry = $this->do_mapping( $form, $entry );

			$target_form_id = $this->target_form_id;

			$new_entry['form_id'] = $target_form_id;

			$target_entry_id = rgar( $entry, $this->update_entry_id );

			$target_entry_id = apply_filters( 'gravityflowformconnector_update_entry_id', $target_entry_id, $target_form_id, $entry, $form, $this );

			if ( empty( $target_entry_id ) ) {
				return true;
			}

			$target_entry = $this->get_remote_entry( $target_entry_id, $target_form_id, $entry, $form );

			if ( $target_form_id !== $target_entry['form_id'] ) {
				$this->log_debug( __METHOD__ . '(): The form of selected entry does not match step settings expected form. No updates performed.' );
				return true;
			}

			switch ( $this->action ) {
				case 'update':
				case 'user_input':
					foreach ( $new_entry as $key => $value ) {
						$target_entry[ (string) $key ] = $value;
					}

					$result = $this->update_remote_entry( $target_entry );

					$this->log_debug( __METHOD__ . '(): update result - ' . print_r( $result, true ) );

					if ( $this->action == 'user_input' ) {
						$assignee_key = strtolower( urlencode( sanitize_text_field( $this->remote_assignee ) ) );
						$route = 'entries/' . $target_entry_id . '/assignees/' . $assignee_key;
						$body  = json_encode( array( 'status' => 'complete' ) );

						$assignee_update_result = $this->remote_request( $route, 'POST', $body );
						$this->log_debug( __METHOD__ . '(): update assignee result - ' . print_r( $assignee_update_result, true ) );
					}

					break;
				case 'approval':
					$assignee_key    = strtolower( urlencode( sanitize_text_field( $this->remote_assignee ) ) );
					$status          = sanitize_text_field( strtolower( rgar( $entry, $this->approval_status_field ) ) );
					$route           = sprintf( 'entries/%d/assignees/%s', $target_entry_id, $assignee_key );
					$body            = json_encode( array( 'status' => $status ) );
					$this->remote_request( $route, 'POST', $body );
			}

			return true;
		}

		/**
		 * Helper to get the entry from the local site based on either a specified entry_id or entry filter criteria.
		 *
		 * @param string $entry_id
		 * @param string $form_id
		 * @param array  $entry
		 * @param array  $form
		 *
		 * @return array
		 */
		public function get_local_entry( $entry_id, $form_id, $entry, $form ) {

			if ( empty( $this->lookup_method ) || $this->lookup_method == 'select_entry_id_field' ) {

				if ( empty( $entry_id ) ) {
					return false;
				}

				$entry_id = apply_filters( 'gravityflowformconnector_target_entry_id', $entry_id, $form_id, $entry, $form, $this );

				$result_entry = GFAPI::get_entry( $entry_id );

			} elseif ( $this->lookup_method == 'filter' ) {

				if ( empty( $this->entry_filter ) ) {

					$this->log_debug( __METHOD__ . '(): No Entry Filter search criteria defined.' );
					return false;

				} else {

					$search_criteria = $this->get_entry_lookup_search_criteria( $form, $entry );

					$sort_criteria = $this->get_entry_lookup_sort_criteria();

					$paging_criteria = array(
						'offset'    => 0,
						'page_size' => 1,
					);

					$entries = GFAPI::get_entries( $form_id, $search_criteria, $sort_criteria, $paging_criteria );

					if ( is_wp_error( $entries ) || empty( $entries ) ) {
						$this->log_debug( __METHOD__ . '(): No entries found that match search criteria.' );
						return false;
					}

					$result_entry = current( $entries );
					$result_entry_id = rgar( $result_entry, 'id' );

					$this->log_debug( __METHOD__ . '(): Filter result is entry #' . $result_entry_id );

					$result_entry_id = apply_filters( 'gravityflowformconnector_target_entry_id', $result_entry_id, $form_id, $entry, $form, $this );

					if ( rgar( $result_entry, 'id' ) != $result_entry_id ) {

						$this->log_debug( __METHOD__ . '(): gravityflowformconnector_target_entry_id filter updated selection to entry #' . $result_entry_id );
						$result_entry = GFAPI::get_entry( $result_entry_id );

					}
				}
			}
			return $result_entry;
		}

		/**
		 * Helper to get the entry from a remote site based on either a specified entry_id or entry filter criteria.
		 *
		 * @param string $entry_id
		 * @param string $form_id
		 * @param array  $entry
		 * @param array  $form
		 *
		 * @return array
		 */
		public function get_remote_entry( $entry_id, $form_id, $entry, $form ) {

			if ( empty( $this->lookup_method ) || $this->lookup_method == 'select_entry_id_field' ) {

				if ( empty( $entry_id ) ) {
					return false;
				}

				$entry_id = apply_filters( 'gravityflowformconnector_target_entry_id', $entry_id, $form_id, $entry, $form, $this );

				if ( empty( $entry_id ) ) {
					return true;
				}

				$route  = 'entries/' . $entry_id;
				$result_entry = $this->remote_request( $route );

			} elseif ( $this->lookup_method == 'filter' ) {

				if ( empty( $this->entry_filter ) ) {

					$this->log_debug( __METHOD__ . '(): No Entry Filter search criteria defined.' );
					return false;

				} else {

					$search_criteria = $this->get_entry_lookup_search_criteria( $form, $entry );

					$sort_criteria = $this->get_entry_lookup_sort_criteria();

					$paging_criteria = array(
						'offset'    => 0,
						'page_size' => 1,
					);

					$route  = 'forms/' . $form_id . '/entries';

					$query_args = array(
						'search' => json_encode( $search_criteria ),
						'sorting' => $sort_criteria,
						'paging' => $paging_criteria,
					);

					$entries = $this->remote_request( $route, 'GET', null, $query_args );

					if ( is_wp_error( $entries ) || empty( $entries ) ) {
						$this->log_debug( __METHOD__ . '(): No entries found that match search criteria.' );
						return false;
					}

					$result_entry = current( $entries['entries'] );

					$result_entry_id = rgar( $result_entry, 'id' );

					$this->log_debug( __METHOD__ . '(): Filter result is entry #' . $result_entry_id );

					$result_entry_id = apply_filters( 'gravityflowformconnector_target_entry_id', $result_entry_id, $form_id, $entry, $form, $this );

					if ( rgar( $result_entry, 'id' ) != $result_entry_id ) {

						$this->log_debug( __METHOD__ . '(): gravityflowformconnector_target_entry_id filter updated selection to entry #' . $result_entry_id );
						$route  = 'entries/' . $result_entry_id;
						$result_entry = $this->remote_request( $route );

					}
				}
			}
			return $result_entry;
		}


		/**
		 * Defines the search criteria for entry when Lookup Conditional Logic has been set in step settings
		 *
		 * @deprecated 1.5.1
		 * @since 1.5.0
		 *
		 * @param array|null $form
		 * @param array|null $entry
		 *
		 * @return array
		 */
		public function gravityflow_entry_lookup_search_criteria( $form = null, $entry = null ) {
			_deprecated_function( __METHOD__, '1.5.1', 'Gravity_Flow_Step_Update_Entry::get_entry_lookup_search_criteria' );

			return $this->get_entry_lookup_search_criteria( $form, $entry );
		}

		/**
		 * Defines the search criteria for entry when Lookup Conditional Logic has been set in step settings
		 *
		 * @since 1.5
		 *
		 * @param array|null $form
		 * @param array|null $entry
		 *
		 * @return array
		 */
		public function get_entry_lookup_search_criteria( $form = null, $entry = null ) {

			$search = array();

			if ( empty( $this->entry_filter ) ) {
				$this->log_debug( __METHOD__ . '(): No Entry Filter search criteria defined.' );
			} else {
				$search['status'] = 'active';
				if ( ! empty( $this->entry_filter['filters'] ) ) {
					$search['field_filters']['mode'] = $this->entry_filter['mode'];
					foreach ( $this->entry_filter['filters'] as $field_filter ) {
						$field_filter_key = $field_filter['field'] == 'entry_id' ? 'id' : $field_filter['field'];

						/* Process merge tags */
						if ( $form !== null && $entry !== null ) {
							$field_filter['value'] = GFCommon::replace_variables( $field_filter['value'], $form, $entry, false, false, false, 'text' );
						}

						$search['field_filters'][] = array(
							'key'      => $field_filter_key,
							'operator' => $field_filter['operator'],
							'value'    => $field_filter['value'],
						);
					}
				}
				$this->log_debug( __METHOD__ . '(): Entry Filter search criteria: ' . print_r( $search, true ) );
			}

			return $search;
		}

		/**
		 * Defines the sort criteria for entry when Lookup Conditional Logic has been set in step settings
		 *
		 * @deprecated 1.5.1
		 * @since 1.5.0
		 *
		 * @return array
		 */
		public function gravityflow_entry_lookup_sort_criteria() {
			_deprecated_function( __METHOD__, '1.5.1', 'Gravity_Flow_Step_Update_Entry::get_entry_lookup_sort_criteria' );

			return $this->get_entry_lookup_sort_criteria();
		}

		/**
		 * Defines the sort criteria for entry when Lookup Conditional Logic has been set in step settings
		 *
		 * @since 1.5.1
		 *
		 * @return array
		 */
		public function get_entry_lookup_sort_criteria() {

			$sort = array();
			if ( ! empty( $this->entry_filtersort_key ) && ! empty( $this->entry_filtersort_direction ) ) {
				$field_sort_key = $this->entry_filtersort_key == 'entry_id' ? 'id' : $this->entry_filtersort_key;

				$sort = array(
					'key' => $field_sort_key,
					'direction' => $this->entry_filtersort_direction,
				);
				$this->log_debug( __METHOD__ . '(): Entry Filter sort criteria: ' . print_r( $sort, true ) );
			} else {
				$this->log_debug( __METHOD__ . '(): No Entry Filter sort criteria defined.' );
			}

			return $sort;
		}


		/**
		 * Updates a remote entry.
		 *
		 * @param $entry
		 *
		 * @return bool
		 */
		public function update_remote_entry( $entry ) {
			$route  = 'entries/' . absint( $entry['id'] );
			$method = 'PUT';
			$body   = json_encode( $entry );

			$result = $this->remote_request( $route, $method, $body );

			return $result;
		}

		/**
		 * Returns the steps for the remote entry.
		 *
		 * @param $form_id
		 *
		 * @return bool
		 */
		public function get_remote_steps( $form_id ) {
			$route = 'forms/' . $form_id . '/steps';
			$steps = $this->remote_request( $route );

			return $steps;
		}

		/**
		 * Returns the remote assignees.
		 *
		 * @param $form_id
		 *
		 * @return array
		 */
		public function get_remote_assignee_choices( $form_id ) {
			$steps         = $this->get_remote_steps( $form_id );
			if ( empty( $steps ) ) {
				return array();
			}
			$assignee_keys = $choices = array();
			foreach ( $steps as $step ) {
				foreach ( $step['assignees'] as $assignee ) {
					$assignee_keys[ $assignee['key'] ] = $assignee['display_name'];
				}
			}

			foreach ( $assignee_keys as $assignee_key => $display_name ) {
				$choices[] = array( 'label' => $display_name, 'value' => $assignee_key );
			}

			return $choices;
		}

		/**
		 * Returns the remote assignees.
		 *
		 * @param $form_id
		 *
		 * @return array
		 */
		public function get_local_assignee_choices( $form_id ) {

			$steps = gravity_flow()->get_steps( $form_id );

			if ( empty( $steps ) ) {
				return array();
			}

			$assignee_keys = $choices = array();

			foreach ( $steps as $step ) {
				$assignees = $step->get_assignees();
				foreach ( $assignees as $assignee ) {
					$assignee_keys[ $assignee->get_key() ] = $assignee->get_display_name();
				}
			}

			$source_form = $this->get_form();

			$choices[] = array(
				'label' => __( 'Select an assignee', 'gravityflow' ),
				'value' => '',
			);

			if ( rgar( $source_form, 'requireLogin' ) ) {
				$choices[] = array(
					'label' => __( 'User (created_by)', 'gravityflow' ),
					'value' => 'created_by',
				);
			}

			foreach ( $assignee_keys as $assignee_key => $display_name ) {
				$choices[] = array( 'label' => $display_name, 'value' => $assignee_key );
			}

			return $choices;
		}
	}
}

Gravity_Flow_Steps::register( new Gravity_Flow_Step_Update_Entry() );
